chpl/chpl-api/src/main/java/gov/healthit/chpl/registration/RateLimitingInterceptor.java

```diff
@@ -0,0 +1,122 @@
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.PreDestroy;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.EnvironmentAware;
+import org.springframework.context.MessageSource;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.context.support.DefaultMessageSourceResolvable;
+import org.springframework.core.env.Environment;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+
+@Component
+public class RateLimitingInterceptor extends HandlerInterceptorAdapter implements EnvironmentAware{
+ 
+    private static final Logger logger = LogManager.getLogger(RateLimitingInterceptor.class);
+    
+    @Autowired
+    private Environment env;
+    
+    @Autowired
+    private MessageSource messageSource;
+    
+    private String timeUnit;
+    
+    private int limit;
+ 
+    private Map<String, SimpleRateLimiter> limiters = new ConcurrentHashMap<>();
+    
+    private List<String> whitelist = new ArrayList<String>();
+    
+    public RateLimitingInterceptor(){}
+    
+    public String getMessage(final String messageCode, final String input, final String input2) {
+        return String.format(messageSource.getMessage(
+                new DefaultMessageSourceResolvable(messageCode),
+                LocaleContextHolder.getLocale()), input, input2);
+    }
+    
+    @Override
+    public void setEnvironment(Environment env) {
+        logger.info("setEnvironment");
+        this.env = env;
+        this.timeUnit = env.getProperty("rateLimitTimeUnit");
+        this.limit = Integer.valueOf(env.getProperty("rateTokenLimit"));
+        this.whitelist = Arrays.asList(env.getProperty("whitelist").split("\\s*,\\s*"));
+    }
+     
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+            throws Exception {
+        String key = null;
+        String clientIdParam = request.getParameter("api_key");
+        String clientIdHeader = request.getHeader("API-Key");
+        
+        if (clientIdHeader == clientIdParam) {
+            key = clientIdHeader;
+        } else {
+            if (clientIdHeader == null) {
+                key = clientIdParam;
+            } else if (clientIdParam == null) {
+                key = clientIdHeader;
+            } else {
+                return false;
+            }
+        }
+        
+        // let non-API requests pass
+        if (whitelist.contains(key)) {
```

Comment:
Does this mean that we have to put the UIs API key in the environment.properties file? I don't like having to have it in both the DB and properties. Is there a way for you to just query it from the DB? Or perhaps add a column to the DB to flag any particular API key as whitelisted?